core/modules/wbase/SendChannelShared.cc

```diff
@@ -0,0 +1,223 @@
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.wbase.SendChannelShared");
+}
+
+namespace lsst {
+namespace qserv {
+namespace wbase {
+
+SendChannelShared::Ptr SendChannelShared::create(SendChannel::Ptr const& sendChannel,
+                                                 wcontrol::TransmitMgr::Ptr const& transmitMgr)  {
+    auto scs = shared_ptr<SendChannelShared>(new SendChannelShared(sendChannel, transmitMgr));
+    return scs;
+}
+
+
+void SendChannelShared::setTaskCount(int taskCount) {
+    _taskCount = taskCount;
+}
+
+
+bool SendChannelShared::transmitTaskLast(StreamGuard sLock, bool inLast) {
+    /// _caller must have locked _streamMutex before calling this.
+    if (not inLast) return false; // This wasn't the last message buffer for this task, so it doesn't matter.
+    ++_lastCount;
+    bool lastTaskDone = _lastCount >= _taskCount;
+    return lastTaskDone;
+}
+
+
+bool SendChannelShared::kill(StreamGuard sLock) {
+    LOGS(_log, LOG_LVL_DEBUG, "SendChannelShared::kill() called");
+    bool ret = _sendChannel->kill();
+    _lastRecvd = true;
+    return ret;
+}
+
+string SendChannelShared::makeIdStr(int qId, int jId) {
+    string str("QID" + (qId == 0 ? "" : to_string(qId) + "#" + to_string(jId)));
+    return str;
+}
+
+
+bool SendChannelShared::addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
+                                    TransmitData::Ptr const& tData, int qId, int jId) {
+    QSERV_LOGCONTEXT_QUERY_JOB(qId, jId);
+    assert(tData != nullptr);
+
+    // This lock may be held for a very long time.
+    std::unique_lock<std::mutex> qLock(_queueMtx);
+    _transmitQueue.push(tData);
+
+    // If _lastRecvd is true, the last message has already been transmitted and
+    // this SendChannel is effectively dead.
+    bool reallyLast = _lastRecvd;
+    string idStr(makeIdStr(qId, jId));
+
+    // If something bad already happened, just give up.
+    if (reallyLast || isDead()) {
+        // If there's been some kind of error, make sure that nothing hangs waiting
+        // for this.
+        LOGS(_log, LOG_LVL_WARN, "addTransmit getting messages after isDead or reallyLast " << idStr);
+        _lastRecvd = true;
+        return false;
+    }
+    {
+        lock_guard<mutex> streamLock(streamMutex);
+        reallyLast = transmitTaskLast(streamLock, last);
+    }
+
+    if (reallyLast || erred || cancelled) {
+        _lastRecvd = true;
+    }
+
+    // If this is reallyLast or at least 2 items are in the queue, the transmit can happen
+    if (reallyLast || _transmitQueue.size() >= 2) {
+        bool scanInteractive = tData->scanInteractive;
+        // If there was an error, give this high priority.
+        if (erred) scanInteractive = true;
+        int czarId = tData->czarId;
+        // Don't wait if cancelled or error.
+        if (erred || cancelled) {
+            return _transmit(erred, largeResult, czarId);
```

Comment:
I simply don't understand the logic here. _transmit() seems hell bent on sending something irrespective of erred or cancelled being set. Even by passing only erred doesn't give it a clue that it might have been cancelled. Why a) pass this at all, b) if you are erred or cancelled why is being called at all? I would think you just want to clean up by quickly deep-sixing the messages.

Reply:
If there was an error, there is probably an error message that needs to sent to the czar. It's probably best if the czar starts retrying this query, or more likely squashing the entire user query asap.  If it was cancelled, the SendChannel is likely dead, so there's probably nothing that can be done. Just finish up asap to free up this this thread and resources.